### PR TITLE
Allow custom filenames for anonymous inputs

### PR DESCRIPTION
**Pull request by mitaa**
This came out of #29253 but doesn't fix it. 

I thought it might be worth merging on its own nonetheless.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I think doctests are a problem here because we call the compiler in slightly weird ways (and don't map the source correctly), but in any case

r? @nrc

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -173,7 +173,10 @@ pub enum Input {
     /// Load source from file
     File(PathBuf),
     /// The string is the source
-    Str(String)
+    Str {
+        input: String,
+        file: String,
```
Comment (nrc) [inline src/librustc/session/config.rs]: These need better names or a more detailed comment

**Comment by nrc** [comment]
lgtm, just needs the naming/documentation improved a bit

**Comment by mitaa** [comment]
(updated)

Does this look better?

**Pull request merged**